arcgisUM.py

```diff
@@ -88,39 +88,84 @@ def getArcGISGroupByTitle(arcGISAdmin, title):
-    
+def addCanvasUsersToGroup(group, courseUsers, instructorLog):
+    """Add new users to a specific ArcGIS group"""
+
     logger.info("addCanvasUsersToGroup: enter")
+    groupNameAndID = util.formatNameAndID(group)
     
     if len(courseUsers) == 0:
         logger.info('No new users to add to ArcGIS Group {}'.format(groupNameAndID))
+        instructorLog += "No new users were added.\n\n"
         return instructorLog
 
     logger.info('Adding Canvas Users to ArcGIS Group {}: {}'.format(groupNameAndID, courseUsers))
     # ArcGIS usernames are U-M uniqnames with the ArcGIS organization name appended.
     arcGISFormatUsers = formatUsersNamesForArcGIS(courseUsers)
     logger.debug("addCanvasUsersToGroup: formatted: {}".format(arcGISFormatUsers))
-    
-    results = group.add_users(arcGISFormatUsers)
-    logger.debug("adding: results: {}".format(results))
 
-    usersNotAdded = results.get('notAdded')
-    """:type usersNotAdded: list"""
     usersCount = len(arcGISFormatUsers)
-    usersCount -= len(usersNotAdded) if usersNotAdded else 0
+    listsOfFormattedUsernames = util.splitListIntoSublists(arcGISFormatUsers, 20)
+    usersNotAdded = []
+
+    for listOfFormattedUsernames in listsOfFormattedUsernames:
+        try:
+            results = group.add_users(listOfFormattedUsernames)
+            logger.debug("adding: results: {}".format(results))
+            usersNotAdded += results.get('notAdded')
+        except RuntimeError as exception:
+            logger.error('Exception while removing users from ArcGIS group "{}": {}'.format(groupNameAndID, exception))
+            return None
+
+    usersCount -= len(usersNotAdded)
     logger.debug("usersCount: {}".format(usersCount))
-    logger.debug("aCUTG: instructorLog 1: [{}]".format(instructorLog))
     instructorLog += 'Number of users added to group: [{}]\n\n'.format(usersCount)
-    logger.debug("aCUTG: instructorLog 2: [{}]".format(instructorLog))
     if usersNotAdded:
         logger.warning('Warning: Some or all users not added to ArcGIS group {}: {}'.format(groupNameAndID, usersNotAdded))
-        instructorLog += 'Users not in group (these users need ArcGIS accounts created for them):\n' + '\n'.join(['* ' + userNotAdded for userNotAdded in usersNotAdded]) + '\n\n' + 'ArcGIS group ID number:\n{}\n\n'.format(group.id)
-    instructorLog += '- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n'
-    logger.debug("aCUTG: instructorLog 3: [{}]".format(instructorLog))
+        instructorLog += 'Users not in group (these users need ArcGIS accounts created for them):\n' + '\n'.join(
+            ['* ' + userNotAdded for userNotAdded in usersNotAdded]
+        ) + '\n'
 
-    logger.info("addCanvasUsersToGroup: instructorLog: [{}]".format(instructorLog))
+    logger.debug("addCanvasUsersToGroup: instructorLog: [{}]".format(instructorLog))
+    return instructorLog
+
+
+def removeCanvasUsersFromGroup(group, groupUsers, instructorLog):
```

Comment:
I think the functions addCanvasUsersToGroup and removeCanvasUsersFromGroup are very similar in the implementation steps. Refactoring into a shared function with a flag to separate add vs remove would work.

e.g. actionCanvasUsersGroup(group, groupUsers, instructorLog, addOrRemove), 
where addOrRemove has String value of "add" or "remove".

Reply:
Yeah, this makes sense. I'll work on that, should be quick.

Reply:
I really, really dislike passing strings around for this purpose. All it takes is a simple misspelling of `addd` or `reomve` to make everything stop working at runtime.  By using functions specific to the action desired, you will get "compile-time" error messages or messages from the IDE that will warn you of problems before the program ever runs.

I'd suggest an approach like using a private function for the general purpose code, to which you pass the `remove` argument a Boolean value:

```python
def _addOrRemove(x, y, z, remove=False):
  # your code here
  pass

def addCanvasUsersToGroup(x, y, z):
  _addOrRemove(x, y, z, remove=False)

def removeCanvasUsersFromGroup(x, y, z):
  _addOrRemove(x, y, z, remove=True)
```

Reply:
I'm on the fence about this. I don't think the misspelling of a string is really that big of an issue, since it's going to appear in the code once (or twice maybe); it's not asking for user input. Also, I think the boolean is somewhat confusing, since `remove=False` does not imply addition. I sometimes wish there was enum function in Python. 

I'm going to implement this using the string approach first, and then we can continue to discuss this. I'm not opposed to going with this approach if enough people think it's superior.

Reply:
As we discussed IRL, I agree that native enums would be very helpful.  Since this is only to be used internally, I won't argue strongly against using a string to specify the mode.  If this were to be part of an API that we would use in our other programs or offer to the public, then I'd be a lot more concerned about passing a mode string.

Reply:
Okay, I went with the string option at this point. I tried to formalize the modes a little bit by creating this `MODIFY_MODES` global. Let me know if you think that works and/or helps at all.

Reply:
Looks good for now!